File: src/Core/Tests/RazorEngine.Core.VB.Tests/RazorEngineHostTestFixture.vb


Imports System.Collections.Generic
Imports System.Dynamic
Imports System.Linq
Imports System.Threading
Imports NUnit.Framework
Imports RazorEngine
Imports RazorEngine.Configuration
Imports RazorEngine.Templating
Imports RazorEngine.Text
Imports RazorEngine.Language
Imports RazorEngine.Tests.TestTypes
Imports RazorEngine.Tests.TestTypes.BaseTypes



''' <summary>
''' Defines a test fixture that provides tests for the <see cref="TemplateBase"/> type.
''' </summary>
<TestFixture> _
Public Class RazorEngineHostTestFixture
#Region "Tests"
    ''' <summary>
    ''' Tests that the <see cref="RazorEngineHost"/> supports the @model directive.
    ''' </summary>
    ''' <remarks>
    ''' As with it's MVC counterpart, we've added support for the @model declaration. This is to enable scenarios
    ''' where the model type might be unknown, but we can pass in an instance of <see cref="object" /> and allow
    ''' the @model directive to switch the model type.
    ''' </remarks>
    <Test> _
    Public Sub RazorEngineHost_SupportsModelSpan_UsingCSharpCodeParser()
        Using service = New TemplateService()
            Const template As String = "@model List<RazorEngine.Tests.TestTypes.Person>" & vbLf & "@Model.Count"
            Const expected As String = "1"

            Dim model = New List(Of Person)() From {New Person() With {.Forename = "Matt", .Age = 27}
                                                   }
            Dim result As String = service.Parse(template, DirectCast(model, Object), Nothing, Nothing)

            Assert.That(result = expected, "Result does not match expected: " & result)
        End Using
    End Sub

    <Test> _
    Public Sub RazorEngineHost_SupportsModelSpan_WithBaseType_NotGeneric_UsingCSharpCodeParser()
        Dim config = New TemplateServiceConfiguration()
        config.BaseTemplateType = GetType(TemplateBase)
        Using service = New TemplateService(config)
            Const template As String = "@model RazorEngine.Tests.TestTypes.Person" & vbLf & "@Model.Forename"
            Const expected As String = "Matt"

            Dim model = New Person() With {.Forename = "Matt"}
            Dim result As String = service.Parse(template, DirectCast(model, Object), Nothing, Nothing)

            Assert.That(result = expected, "Result does not match expected: " & result)
        End Using
    End Sub

    ''' <summary>
    ''' Tests that the <see cref="RazorEngineHost"/> supports the @ModelType directive.
    ''' </summary>
    ''' <remarks>
    ''' As with it's MVC counterpart, we've added support for the @ModelType declaration. This is to enable scenarios
    ''' where the model type might be unknown, but we can pass in an instance of <see cref="object" /> and allow
    ''' the @model directive to switch the model type.
    ''' </remarks>
    <Test> _
    Public Sub RazorEngineHost_SupportsModelSpan_UsingVBCodeParser()
        Dim config = New TemplateServiceConfiguration() With {.Language = Language.VisualBasic}

        Using service = New TemplateService(config)
            Const template As String = "@ModelType List(Of RazorEngine.Tests.TestTypes.Person)" & vbLf & "@Model.Count"
            Const expected As String = "1"

            Dim model = New List(Of Person)() From {
                New Person() With {
                    .Forename = "Matt",
                    .Age = 27
                }
            }

            Dim result As String = service.Parse(template, DirectCast(model, Object), Nothing, Nothing)

            Assert.That(result = expected, "Result does not match expected: " & result)
        End Using
    End Sub

    <Test> _
    Public Sub RazorEngineHost_SupportsModelSpan_WithBaseType_NotGeneric_UsingVBCodeParser()
        Dim config = New TemplateServiceConfiguration()
        config.BaseTemplateType = GetType(TemplateBase)
        config.Language = Language.VisualBasic

        Using service = New TemplateService(config)
            Const template As String = "@ModelType List(Of RazorEngine.Tests.TestTypes.Person)" & vbLf & "@Model.Count"
            Const expected As String = "1"

            Dim model = New List(Of Person)() From {
                New Person() With {
                    .Forename = "Matt",
                    .Age = 27
                }
            }

            Dim result As String = service.Parse(template, DirectCast(model, Object), Nothing, Nothing)

            Assert.That(result = expected, "Result does not match expected: " & result)
        End Using
    End Sub
#End Region
End Class

File: src/test/Test.RazorEngine.Core/RazorEngineHostTestFixture.cs
using RazorEngine.Tests.TestTypes.BaseTypes;

namespace RazorEngine.Tests
{
    using System.Collections.Generic;
    using System.Dynamic;
    using System.IO;
    using System.Linq;

    using NUnit.Framework;

    using Compilation;
    using Configuration;
    using Templating;
    using Text;
    using TestTypes;

    /// <summary>
    /// Defines a test fixture that provides tests for the <see cref="TemplateBase"/> type.
    /// </summary>
    [TestFixture]
    public class RazorEngineHostTestFixture
    {
        #region Tests
        /// <summary>
        /// Tests that the <see cref="RazorEngineHost"/> supports the @model directive.
        /// </summary>
        /// <remarks>
        /// As with it's MVC counterpart, we've added support for the @model declaration. This is to enable scenarios
        /// where the model type might be unknown, but we can pass in an instance of <see cref="object" /> and allow
        /// the @model directive to switch the model type.
        /// </remarks>
        [Test]
        public void RazorEngineHost_SupportsModelSpan_UsingCSharpCodeParser()
        {
            using (var service = RazorEngineService.Create())
            using (var writer = new StringWriter())
            {
                const string template = "@model List<RazorEngine.Tests.TestTypes.Person>\n@Model.Count";
                const string expected = "1";

                var model = new List<Person> { new Person() { Forename = "Matt", Age = 27 } };
                var key = service.GetKey(nameof(template));

                service.AddTemplate(key, template);
                service.RunCompile(key, writer, modelType: model.GetType(), model: model);

                string result = writer.ToString();

                Assert.That(result == expected, "Result does not match expected: " + result);
            }
        }

#if !RAZOR4
        /// <summary>
        /// Tests that the <see cref="RazorEngineHost"/> supports the @ModelType directive.
        /// </summary>
        /// <remarks>
        /// As with it's MVC counterpart, we've added support for the @ModelType declaration. This is to enable scenarios
        /// where the model type might be unknown, but we can pass in an instance of <see cref="object" /> and allow
        /// the @model directive to switch the model type.
        /// </remarks>
        [Category("VBNET")]
        [Test]
        public void RazorEngineHost_SupportsModelSpan_UsingVBCodeParser()
        {
            var config = new TemplateServiceConfiguration
                             {
                                 Language = Language.VisualBasic
                             };

            using (var service = RazorEngineService.Create(config))
            {
                const string template = "@ModelType List(Of RazorEngine.Tests.TestTypes.Person)\n@Model.Count";
                const string expected = "1";

                var model = new List<Person> { new Person() { Forename = "Matt", Age = 27 } };
                var key = service.GetKey("template");

                service.AddTemplate(key, new LoadedTemplateSource(template));

                string result = service.RunCompile(key, modelType: typeof(object), model: model);

                Assert.That(result == expected, "Result does not match expected: " + result);
            }
        }
#endif
        #endregion
    }
}